fusequery/query/src/catalogs/impls/database_catalog.rs

```diff
@@ -98,106 +110,96 @@ impl DatabaseCatalog {
         table_version: Option<MetaVersion>,
     ) -> Result<Arc<TableMeta>> {
-        let db_lock = self.databases.read();
-        let database = db_lock.get(db_name).ok_or_else(|| {
-            ErrorCode::UnknownDatabase(format!("Unknown database: '{}'", db_name))
-        })?;
-
+        let database = self.get_database(db_name)?;
         let table = database.get_table_by_id(table_id, table_version)?;
         Ok(table.clone())
     }
 
-    pub async fn get_remote_table(
-        &self,
-        db_name: &str,
-        table_name: &str,
-    ) -> Result<Arc<TableMeta>> {
-        match self.get_table(db_name, table_name) {
-            Ok(t) if t.datasource().is_local() => Err(ErrorCode::LogicalError(format!(
-                "local table {}.{} exists, which is used as remote",
-                db_name, table_name
-            ))),
-            tbl @ Ok(_) => tbl,
-            _ => {
-                let cli_provider = self.remote_factory.store_client_provider();
-                let mut store_cli = cli_provider.try_get_client().await?;
-                let res = store_cli
-                    .get_table(db_name.to_string(), table_name.to_string())
-                    .await?;
-                let remote_table = RemoteTable::try_create(
-                    db_name.to_string(),
-                    table_name.to_string(),
-                    res.schema,
-                    self.remote_factory.store_client_provider().clone(),
-                    TableOptions::new(),
-                )?;
-
-                // Remote_table we've got here is NOT cached.
-                //
-                // Since we should solve the metadata synchronization problem in a more reasonable way,
-                // let's postpone it until we have taken all the things into account.
-                Ok(Arc::new(TableMeta::new(
-                    Arc::from(remote_table),
-                    res.table_id,
-                )))
-            }
-        }
-    }
-
-    pub fn get_all_tables(&self) -> Result<Vec<(String, Arc<TableMeta>)>> {
+    fn get_all_tables(&self) -> Result<Vec<(String, Arc<TableMeta>)>> {
         let mut results = vec![];
-        for (k, v) in self.databases.read().iter() {
+        let mut db_names = HashSet::new();
+        for (db_name, v) in self.databases.read().iter() {
             let tables = v.get_tables()?;
             for table in tables {
-                results.push((k.clone(), table.clone()));
+                results.push((db_name.clone(), table.clone()));
             }
+            db_names.insert(db_name.clone());
         }
+
+        if !self.disable_remote {
+            let mut remotes = self
+                .meta_store_cli
+                .get_all_tables()?
+                .into_iter()
+                // local and system dbs should shadow remote db
+                .filter(|(n, _)| !db_names.contains(n))
+                .collect::<Vec<_>>();
+            results.append(&mut remotes);
+        }
```

Comment:
Yes, local tables could be quite different between Query nodes; it is hard to keep them consistent (or we should not be aiming that).

If `get_all_tables` are being concerned with, the current implementation of server-side is packing all the lastest table metas, and returns them back, which is a snapshot of the given point of time.

Reply:
I mean, it only returns tables that belongs to a database that is in `self.databases`, right?
But there may be some database not in `self.databases`. Is that possible?

Reply:
`get_all_tables`  is supposed to return all the tables,  including local and remote databases.

In case of same databases ( with identical names) exist both in local database and remote database, local database will shadow the remote one.

note: while getting database from catalog, there are no args to indicate which type(local/remote) is prefered.

For example, QueryNode_1 may create a **local** database `db1` 
~~~
CREATE DATABASE db1 ENGINE = Local
~~~

after that,  QueryNode_2 may create a **remote** database `db1`
~~~
CREATE DATABASE db1
~~~

and then,  when Q_1 invokes `get_all_tables`, remote db1 and its tables are supposed to be invisible to Q1.

That's basically the quoted code trying to do. And yes, a "local database" is somehow award; maybe Catalog-Plugin of Spark is a better abstraction, I am not quite sure yet, but hoping we can eliminate the local database eventually.

Reply:
What if:
- Q1 creates remote db: db1
- Q1 caches db1
- Q2 creates remote db: db2
- Q2 creates remote table db2.t2
- Q1 `get_all_tables()`. Then Q1 only caches tables belonging to db1;
There is a `db2` but before cache is evicted, Q1 will never see `db2.t2`. Will this happen?

Reply:
No. Remote databases are not cached.

Reply:
My mistake. I was supposing `self.databases` is a set of all local and remote dbs.